File: core/src/main/java/heast/core/network/packets/s2c/login/HelloS2CPacket.java
package heast.core.network.packets.s2c.login;

import heast.core.network.Packet;
import heast.core.network.PacketBuf;
import heast.core.network.listeners.ClientLoginListener;

public final class HelloS2CPacket implements Packet<ClientLoginListener> {

    final byte[] key;

    public HelloS2CPacket(byte[] key) {
        this.key = key;
    }

    public HelloS2CPacket(PacketBuf buf) {
        this.key = buf.readByteArray();
    }

    @Override
    public void write(PacketBuf buf) {
        buf.writeByteArray(key);
    }

    @Override
    public void apply(ClientLoginListener listener) {
        listener.onHello(this);
    }

    public byte[] getKey() {
        return key;
    }
}

File: Core/network/packets/s2c/HelloS2CPacket.cs
using Core.network.listeners;

namespace Core.network.packets.s2c; 

public class HelloS2CPacket : IPacket<IClientLoginListener> {
    
    public byte[] Key { get; }
    
    public HelloS2CPacket(byte[] key) {
        Key = key;
    }
    
    public HelloS2CPacket(PacketBuf buf) {
        Key = buf.ReadByteArray();
    }
    
    public void Write(PacketBuf buf) {
        buf.WriteByteArray(Key);
    }

    public void Apply(IClientLoginListener listener) {
        listener.OnHello(this);
    }
}